internal/utils/cli/error.go

```diff
@@ -0,0 +1,138 @@
+package cli
+
+import (
+	"fmt"
+)
+
+// ErrUsageHider may hide the command's usage when an error occurs
+type ErrUsageHider interface {
+	HideUsage() bool
+}
+
+// ErrSuggester provides a list of suggestions that will display to the user when an error occurs
+type ErrSuggester interface {
+	Suggestions() []interface{}
+}
+
+// ErrLinkReferrer provides a list of reference links that will display to the user when an error occurs
+type ErrLinkReferrer interface {
+	ReferenceLinks() []interface{}
+}
+
+// NewErr returns a new CLI error
+func NewErr(cause error, details ...ErrDetail) error {
+	var d ErrDetails
+	for _, detail := range details {
+		detail.Apply(&d)
+	}
+
+	var hideUsage bool
+	if d.HideUsage != nil {
+		hideUsage = *d.HideUsage
+	}
+
+	return cliErr{
+		cause:          cause,
+		hideUsage:      hideUsage,
+		suggestions:    d.Suggestions,
+		referenceLinks: d.ReferenceLinks,
+	}
+}
+
+// WrapErr wraps the provided error with the provided message
+// The extra details are applied on top of any existing, wrapped details
+// Note: it is assumed the message contains a single `fmt` verb to include the cause/error
+func WrapErr(msg string, cause error, details ...ErrDetail) error {
+	err, ok := cause.(cliErr)
```

Comment:
q: Is there a strong reason why we need `cliError` and `ErrDetails`? Is `ErrDetails` a struct that should only be used as the arguments to the `Apply` interface method?

In the end, I'm fine with keeping both, but wanted to see if we can DRY things up a bit here.

Reply:
`cliError` is the internal representation these helpers are relying on as an "error type" (read: implement `Error() string`)

`ErrDetails` is the "input struct" presenting calling code with the various options present to inject into the error.  it's fields are more options (`*bool` vs `bool`) and doesn't implement the `error` interface